File: core/src/main/java/com/emarsys/core/database/CoreSQLiteDatabase.java
package com.emarsys.core.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.emarsys.core.database.trigger.TriggerEvent;
import com.emarsys.core.database.trigger.TriggerType;

public interface CoreSQLiteDatabase {

    SQLiteDatabase getBackingDatabase();

    Cursor rawQuery(String sql, String[] selectionArgs);

    Cursor query(
            boolean distinct,
            String table,
            String[] columns,
            String selection,
            String[] selectionArgs,
            String groupBy,
            String having,
            String orderBy,
            String limit);

    void execSQL(String sql);

    long insert(String table, String nullColumnHack, ContentValues values);

    int update(String table, ContentValues values, String whereClause, String[] whereArgs);

    int delete(String table, String whereClause, String[] whereArgs);

    void registerTrigger(
            String table,
            TriggerType triggerType,
            TriggerEvent triggerEvent,
            Runnable trigger);

    void beginTransaction();

    void setTransactionSuccessful();

    void endTransaction();

}

File: core/src/main/java/com/emarsys/core/database/CoreSQLiteDatabase.kt
package com.emarsys.core.database

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.emarsys.core.database.trigger.TriggerEvent
import com.emarsys.core.database.trigger.TriggerType

interface CoreSQLiteDatabase {
    val backingDatabase: SQLiteDatabase?
    fun rawQuery(sql: String, selectionArgs: Array<String>?): Cursor
    fun query(
            distinct: Boolean,
            table: String,
            columns: Array<String>?,
            selection: String?,
            selectionArgs: Array<String>?,
            groupBy: String?,
            having: String?,
            orderBy: String?,
            limit: String?): Cursor

    fun execSQL(sql: String)
    fun insert(table: String, nullColumnHack: String?, values: ContentValues): Long
    fun update(table: String, values: ContentValues, whereClause: String?, whereArgs: Array<String>?): Int
    fun delete(table: String, whereClause: String?, whereArgs: Array<String>?): Int
    fun registerTrigger(
            table: String,
            triggerType: TriggerType,
            triggerEvent: TriggerEvent,
            trigger: Runnable)

    fun beginTransaction()
    fun setTransactionSuccessful()
    fun endTransaction()
}